samza-core/src/main/java/org/apache/samza/runtime/ContainerLaunchUtil.java

```diff
@@ -142,19 +150,59 @@ public void beforeStart() {
             public void afterStart() {
               log.info("Container Started");
               listener.afterStart();
+              addShutdownHook();
             }
 
             @Override
             public void afterStop() {
               log.info("Container Stopped");
               listener.afterStop();
+              removeShutdownHook();
+              shutdownLatch.countDown();
```

Comment:
Can you please double check that either `afterStop` or `afterFailure` will always be called, even in failure cases? Ideally, the shutdown timeout is only hit because some part of the shutdown was taking a long time, but not because there was some flow which accidentally forgot to call the callback.
I think you are good here, but just double checking.

Reply:
Look like either `afterStop` or `afterFailure` will always be called.
On graceful shutdown (via container.shutdown() or runLoop shutting down) -> `afterStop`
On Exception -> `afterFailure`